[1단계 - 출석 구현] 부기(이창의) 미션 제출합니다.

changuii opened the pull request:
## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?

[부기 프롤로그 링크](https://prolog.techcourse.co.kr/studylogs/3953)


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
시간이 많이 부족했고, 구현을 가장 최우선 목표로 작성하다보니 코드를 객체지향적으로 많이 작성하지 못했습니다. 리팩터링할 시간이 부족했습니다.
  
## 어떤 부분에 집중하여 리뷰해야 할까요?
각 클래스들이 상호작용하는 부분(클래스들의 책임)에 대해서 집중해서 리뷰부탁드리겠습니다.!

phk1128 committed:
docs(README): 기능 요구 사항 작성 하라

- 파일 입출력 기능 요구 사항 작성

phk1128 committed:
test(FileManagerTest): csv 데이터 입출력 레드 테스트하라

- csv 파일 전체를 읽어온다

phk1128 committed:
test(FileManagerTest): csv 파일 입출력을 그린 테스트하라

- csv 파일 전체를 읽어온다.

phk1128 committed:
feat(FileManager): csv 파일을 읽어와라

- readFileLines() 메서드의 프로덕션 코드를 작성하였다.

phk1128 committed:
docs(README): 완성된 파일 입출력 기능을 체크하라

- 리소스에 있는 csv 파일에 데이터 입력 기능을 완성하였다.

phk1128 committed:
feat(CrewName): 크루 이름 도메인을 구현하라

phk1128 committed:
feat(Attendance): 출석 도메인을 구현하라

- Attendance는 출석을 담당하는 도메인 객체
- AttendanceStatus는 출석의 상태를 나타내는 enum 객체

phk1128 committed:
feat(Crew): 크루 도메인을 구현하라

- ExpulsionStatus는 제적 위험 상태를 나타내는 객체

phk1128 committed:
feat: 크루 출석 기록 확인 기능을 구현하라

phk1128 committed:
feat : 크루의 출석 확인 기능을 구현하라

phk1128 committed:
feat : 날짜에 대한, 새로운 Attendance 수정 기능을 구현하라

phk1128 committed:
feat: 제적 위험자 확인 기능을 구현하라

phk1128 committed:
feat: controller에서 각각의 기능들을 조합하라

phk1128 committed:
fix: InputView에서, 파싱에 관한 예외를 IllegalArgumentException으로 던져라

phk1128 committed:
test: Crew 도메인에 대한 단위 테스트를 작성하라

phk1128 committed:
feat: 이미 출석이 존재한다면, 수정으로 안내하라

changuii committed:
test: AttendanceSystem에 대한 단위 테스트를 작성하라

changuii committed:
fix: 클래스파일을 git에서 제거하라

phk1128 committed:
test(AttendanceTest): Nested로 성공 테스트를 분리하라

phk1128 committed:
refactor: 인덴트가 2인 부분을 1로 수정하라

phk1128 committed:
refactor: 등교일이 아닌 경우 사용자에게 안내 후 초기 화면으로 돌아가라

phk1128 committed:
docs(README): 기능 구현 목록을 체크하라

Rok93 commented on src/main/resources/attendances.csv:
```diff
@@ -0,0 +1,42 @@
+nickname,datetime
+쿠키,2024-12-13 10:08
```
여기 값은 부기가 직접 입력하신 값인가요? 
아니면 따로 미션을 위해 전달받은 파일이 있는걸까요?

크루마다 조금씩 다른 것 같아서 질문드려요! 👀

changuii replied:
이번 미션에서는 주어진 파일이 없고 저희는 최종 코딩테스트에서 주신 파일로 했어요!

Rok93 replied:
엇... 최종 코딩테스트에서 주신 파일로 했다면 크루들이 다 같아야하는게 아닌가 싶은데, 크루들마다 다른 조금씩 다른 방식으로 파일을 쓰셨나보네요. 🤔

changuii replied:
네 맞아요 알아서 하라고 하셨어요 🤣

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+
+    private final InputView inputView;
+    private final OutputView outputView;
+    private final AttendanceSystem attendanceSystem;
```
Controller가 마치 도메인 객체러럼 `AttendanceSystem`(도메인 객체)를 상태를 가지고 있군요? 🤔
Controller도 도메인 객체라고 볼 수 있을까요?

changuii replied:
controller가 도메인 객체라고 생각하기 보다는, 하나의 프로그램의 시작부터 종료까지의 사이클동안 `AttendanceSystem` 도메인 객체가 하나만 만들어지기에, controller가 생성될 때, 출석 정보들을 생성해두고 사용한다라는 느낌으로 작성했습니다.
제가 생각하기에는 해당 프로그램에서 문제가 없다고 생각합니다.

- controller는 상태를 가지면 안되는걸까요? 
- 만약 여러 사용자가 있는 경우에 데이터의 정합성 문제 때문에 controller가 상태를 가지면 안되는 걸까요?
- 아니면 AttendanceSystem이 run()이라는 하나의 메서드 내에서만 사용되기 때문일까요? 

저는 해당 프로그램이 csv파일에 데이터를 업데이트 하지 않고, csv파일을 읽어서 실행한 후 client가 controller를 삭제하는 순간까지 해당 정보를 가지고 있다고 판단했는데,
로키는 어떻게 생각하시는지 궁금합니다. 🤔

Rok93 replied:
먼저 Controller의 역할을 생각해봤을 때, AttendanceSystem을 상태로 가질 필요가 없다고 생각했어요.

사실 지금은 싱글 스레드로 동작할 것이기 때문에 동시에 여러 요청을 처리할 일이 없겠지만, 추후 조금 더 확장하여 controller가 여러 사용자에게 동시에 호출되게되면 이 `AttendanceSystem` 전역 상태처럼 사용될 여지가 생길 수 있을 것 같아요.

제가 우려하는건 `만약 여러 사용자가 있는 경우에 데이터의 정합성 문제 때문에 controller가 상태를 가지면 안되는 걸까요?` 에 가깝긴 하네요. 😃 

controller가 꼭 AttendanceSystem를 상태로 가지고있을 필요성을 크게 못느끼겠어요. 🤔
단순히 run 하는 시점에 객체를 생성하여 로컬 변수로 사용해도 충분할 것 같은데, 굳이 상태로 가질 필요가 없다고 생각되기도하네요.

changuii replied:
> 사실 지금은 싱글 스레드로 동작할 것이기 때문에 동시에 여러 요청을 처리할 일이 없겠지만, 추후 조금 더 확장하여 controller가 여러 사용자에게 동시에 호출되게되면 이 AttendanceSystem 전역 상태처럼 사용될 여지가 생길 수 있을 것 같아요.

저도 이 부분에 대해서 동의합니다. 하지만, 요구사항으로만 판단했을때는 사용자가 늘어나는 경우가 없다고 판단했습니다. 이런 판단은 어떤 기준으로 내려야 할까요.. 
만약 웹 애플리케이션과 같이 AttendanceSystem에 여러 사용자가 접근하는 경우에는 레이스 컨디션을 고려를 해야겠네요!  

> controller가 꼭 AttendanceSystem를 상태로 가지고있을 필요성을 크게 못느끼겠어요. 🤔

이 부분에 대해서 동의합니다.!! 저도 굳이 가질 필요는 없다고 생각되지만, 이 부분에 대해서 "굳이"라는 용어가 너무 감에 의존하거나 모호해서 다른 분들에게 설명할때 사실 어떻게 설명해야할지 잘 모르겠습니다.
굳이 상태로 가질 필요가 없다고 생각되시는 이유가 객체의 필드가 높아질수록 오류가 많아지기 때문인가요 ? 🤔

Rok93 replied:
일반적으로 인스턴스 변수는 가급적 줄이는 것이 좋아요. 
객체가 굳이 필요하지 않은 상태 값을 일일이 상태로 가지게되면 객체가 주입받아야하는 상태 값도 많아지기 때문에 비대해지게되고 
많은 상태값들을 갖다보니 할 수 있는 것들이 많아지니 많은 책임을 가지게되고 자연스럽게 그 객체는 비대해지게될거에요. 

물론 객체를 생성하는데 많은 상태값을 주입해줘야하니 테스트도 어려워지구요. 🥲

그렇기 때문에 인스턴스 변수를 굳이 둘 필요가 없다면 제거하는 쪽이 훨씬 좋다고 생각해요.

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+        outputView.printToday(LocalDate.now().withYear(2024).withMonth(12));
+        outputView.printIntroduceOperation();
+        final Operation operation = inputView.readChoiceOperation();
+        if (operation == Operation.ADD_ATTENDANCE) {
```
Operation도 나쁘지 않지만 Command 와 같은 표현을 써봐도 좋을 것 같아요.

changuii replied:
감사합니다. 주로 이런 경우에 관용적으로 사용되는 네이밍이 `Command` 일까요?

Rok93 replied:
디자인 패턴 중에서도 [커맨드 패턴](https://gsbang.tistory.com/entry/Design-Pattern-Command-Pattern%EC%BB%A4%EB%A7%A8%EB%93%9C-%ED%8C%A8%ED%84%B4)이라는 케이스가 있어요. 😃 
이에 착안에서 제안드렸던거였어요.  관용적으로 쓰이냐에 대해서는 확신을 가지고 말씀드리기는 어렵지만, 저는 command 라는 용어를 꽤나 자주 접했고 자주 사용하고있기는합니다. 🤔

Operation이라는 용어를 직접 접하지는 못했지만 Operator의 경우에도 많이 봤던 네이밍이라 사용하신 워딩도 괜찮은 네이밍이라 생각하지만, 개인적으로는 Command가 `명령`, `지시`의 개념의 표현이라 조금 더 적절하다고 생각했어요.

<img width="280" alt="image" src="https://github.com/user-attachments/assets/346c6e37-b4f0-4ac3-a5c7-efc3d522ce3e" />

changuii replied:
감사합니다. 반영하겠습니다!!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+    }
+
+
+    public Operation selectOperation() {
```
selectOperation 메서드는 외부에서 사용하지 않는 것 같은데요. 접근제한자를 `private`로 제한해주면 좋을 것 같아요.

changuii replied:
빠뜨린 것 같습니다. 리뷰 감사합니다 :)

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+        while (selectOperation() != Operation.QUIT) {
+        }
+    }
+
```
불필요한 줄바꿈은 제거해주세요.  🙏
```suggestion
```

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+        if (answerCommand == AnswerCommand.YES) {
+            final int dayOfMonth = LocalDate.now().getDayOfMonth();
+            final LocalTime targetTime = LoopTemplate.tryCatchLoop(this::inputUpdateTime, outputView);
+            final Attendance beforeAttendance = attendanceSystem.findAttendanceByDate(crewName, dayOfMonth);
+            final Attendance afterAttendance = attendanceSystem.updateAttendanceByCrewNameAndDay(targetTime, crewName,
```
아래와 같이 찾아온 attendance 객체에게 출석 시간을 수정하라고 메시지를 보내보면 되지 않을까요? 😃 
```suggestion
            final Attendance attendance = attendanceSystem.findAttendanceByDate(crewName, dayOfMonth);
            attendance.updateAttendanceTime(targetTime);
```

changuii replied:
> controller가 attendanceSystem에게 해당 크루의 출석을 변경시켜라 -> attendanceSystem은 크루를 찾아서 해당 날짜의 출석을 변경하라 -> 크루는 자신의 출석을 변경한다
위의 과정이 어색한걸까요?

find와 update가 두 번 이뤄진 것은, before와 after의 정보가 모두 필요했기 때문인데, 제가 잘못이해한걸까요?
로키의 생각이 궁금합니다.

Rok93 replied:
아하 ~  왜 부기가 이렇게 작성을 하신건지 이제 이해가 되네요. 

저는 after의 정보만 알면 된다고 생각했었는데, 요구사항에 따르면 before의 정보도 필요하기 때문이었겠군요? 😃

이 부분은 제가 제안드린 방식에서 `updateAttendanceTime` 의 반환 값을 전달해주면 될 것 같은데요. 
예를들자면 `AttendanceUpdateSnapshot` 같은 타입을 반환하여 필요한 before, after의 정보를 가지도록해볼 수도 있을 것 같은데, 이렇게한다면 저의 의도와 부기의 의도가 모두 만족되는게 맞을까요?

changuii replied:
말씀해주신 것 처럼 비즈니스 로직을 controller에서 도메인 객체를 제어해서 여러 단계로 수행하고 있습니다. 따라서 객체 내부에서 수행하는 것이 아닌 외부 controller에서 수행하고 있어 모호하긴 합니다. 
다만, 말씀하신 부분에서 before과 after를 묶은 `AttendanceUpdateSnapshot`을 만드신다고 하셨는데 이 경우 객체가 상태는 가지지만 행위는 가지지 않을 것 같습니다.
그럼 dto 처럼 데이터를 전달하기 위한 수단으로 사용하는데 도메인 객체가 dto를 만들어 반환하는게 어색한 것 같습니다.

이에 대한 로키의 생각이 궁금합니다! 🤔

Rok93 replied:
지금의 상황에서는 제가 말한대로라면 DTO 처럼 보일 수 있을 것 같아요.  😢
왜냐하면 적절히 위임해줄 역할이 당장 떠오르지 않기 때문에 그냥 단순히 필요한 데이터만 가진 객체가 되기 때문이죠. 
모든 객체가 역할과 책임을 가지면 객체지향적인 설계를 잘 한 것 같아 기분이 좋겠지만, 사실 모든 객체가 이상적으로 역할, 책임을 가지는게 생각보다 쉽지는 않다고 생각해요. 😅

하지만 지금처럼 before, after의 스냅샷을 Controller에서 각각 남겨서 비즈니스 로직을 수행하는 것 보다는 
 Attendance 객체가 직접 이 두 값을 기록해주는 것이 좋다고 생각해요. 🤔

changuii replied:
아하 dto처럼 보일 수 있지만, "스냅샷을 기록한다."라는 비즈니스 로직을 수행하기 위해 dto 처럼 보이는 객체를 기록해서 반환하는 것이군요 !!

>  사실 모든 객체가 이상적으로 역할, 책임을 가지는게 생각보다 쉽지는 않다고 생각해요. 😅

이것때문에 객체지향이 어려운 것 같아요.. 감사합니다!!

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,180 @@
+    public AttendanceController(final InputView inputView, final OutputView outputView) {
+        this.inputView = inputView;
+        this.outputView = outputView;
+        final List<String> attendanceLines = FileManager.readFileLines(ATTENDANCE_HISTORY_FILE_NAME);
+        attendanceLines.remove(0);
+        attendanceSystem = AttendanceSystem.of(attendanceLines, LocalDate.now().withYear(2024).withMonth(12));
```
(csv 파일을 읽어) `AttendanceSystem` 객체를 생성하는 Factory 클래스를 하나 만들어서 역할을 위임해보면 어떨까요? 😃

changuii replied:
controller의 책임이 많다고 생각하시는 걸까요?
Controller가 생성될 때, AttendanceSystem이 생성된다는 느낌으로 만들었는데 책임이 너무 많은 걸까요? 

책임의 분리가 언제 이뤄져야 할까요? 로키의 생각이 궁금해요!

Rok93 replied:
> controller의 책임이 많다고 생각하시는 걸까요?

Controller의 관심사가 아닌 것 같아서 제안드렸었는데요.
csv 파일을 읽어서 AttendanceSystem 객체를 생성하는 것이 Controller의 역할로 보이지 않았고, 다른 객체에게 이 역할을 위임한다면 테스트 코드를 작성하여 해당 기능을 안전하게 보호할 수도 있을 것 같아서 제안드렸었어요. 😃 

음...  조금 더 생각해보니 csv 파일을 읽어 AttendanceSystem 객체를 생성하는 로직을 Controller 내부에 별도 private 메서드 정도로만 분리해도 충분할 것 같기도하네요. 🤔 

```java
        public AttendanceSystem createAttendanceSystem() {
            final List<String> attendanceLines = FileManager.readFileLines(ATTENDANCE_HISTORY_FILE_NAME);
            attendanceLines.remove(0);
            return AttendanceSystem.of(attendanceLines, LocalDate.of(2025, 12, 20));
        }
```

changuii replied:
> 다른 객체에게 이 역할을 위임한다면 테스트 코드를 작성하여 해당 기능을 안전하게 보호할 수도 있을 것 같아서 제안드렸었어요. 😃

이 부분을 읽고 왜 도메인 생성에 대한 로직을 외부로 노출시키라는지 이해했습니다!
특정 도메인 객체를 생성하는 로직도 중요한 비즈니스 로직이니 테스트하지 못하는 부분이 아닌 외부로 분리하여 테스트할 수 있도록 하라는 말씀이시군요!
- 지난 로또에서는 랜덤 넘버를 생성하는 로직을 외부로 추출해서 단위 테스트를 수월하게 했는데 이 같은 경우에는 단위 테스트를 할 수 없는 위치에 있는 로직을 외부로 분리하는 과정이네요

감사합니다 :)

Rok93 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,40 @@
+    ABSENCE("결석", 30);
+
+    String name;
+    int boundary;
```
접근제한자를 default로 두신 이유가 있을까요? 
또한 enum class의 상태 값들은 상태가 한번 지정되면 변경되지 않을테니 `final`로 선언해주면 좋을 것 같습니다. 

```suggestion
    private final String name;
    private final int boundary;
```

changuii replied:
> 접근제한자를 default로 두신 이유가 있을까요? 

급하다 보니, 빠뜨린 것 같습니다. 

> enum class의 상태 값들은 상태가 한번 지정되면 변경되지 않을테니 final로 선언해주면 좋을 것 같습니다.

enum 내부 필드는 상수 처리 되어있는 줄 알았는데, 말씀해주셔서 테스트 해보니 변경이 되더라구요. 리뷰 감사합니다 !!

Rok93 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,40 @@
+    LATE("지각", 5),
+    ABSENCE("결석", 30);
+
+    String name;
+    int boundary;
```
직관적으로 이해가 안되는데 `boundary` 가 어떤 의미로 네이밍하신건지 궁금합니다.

changuii replied:
기준으로 잡힌 시간(분)에서 어느 지점부터 지각, 결석인지를 나타내는 경계입니다.
ex) 입실이 10시라면, 5는 10시 5분 30은 10시 30분입니다.

너무 간결하게 작성한 걸까요? 아니면 boundary 용어가 잘못된 걸까요?
 요구사항을 읽는다면 이해할 수 있다고 생각했는데 로키의 생각이 궁금합니다. 🤔

Rok93 replied:
새로운 팀원이 팀에 합류하게되면 모든 요구사항을 다 읽어보고 코드를 읽어볼 수 있는게 아닐거에요. 😃 
이 코드를 읽는 사람이 당연히 요구사항을 읽어봤을거라고 가정하면 안될 것 같아요! 
그렇기 때문에 가능하면 코드를 읽었을 때도 어느정도 내용을 파악할 수 있도록 짜는 것이 좋다고 생각합니다.

changuii replied:
> 새로운 팀원이 팀에 합류하게되면 모든 요구사항을 다 읽어보고 코드를 읽어볼 수 있는게 아닐거에요. 😃
이 코드를 읽는 사람이 당연히 요구사항을 읽어봤을거라고 가정하면 안될 것 같아요!

이해가 바로 되었습니다. 리뷰 감사합니다.  😊

Rok93 commented on src/main/java/domain/AttendanceStatus.java:
```diff
@@ -0,0 +1,40 @@
+                .withHour(boundaryHour)
+                .withMinute(boundaryMinute);
+        return Arrays.stream(values())
+                .sorted((o1, o2) -> o2.boundary - o1.boundary)
```
이 정렬은 로직은 아래의 요구사항을 만족하기 위함이실까요? 🤔

> 제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다. 

만약 이 정렬 순서가 바뀌게되면 어떻게될까요? 👀

changuii replied:
> 제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다.

위의 요구사항을 만족하기 위함은 아닙니다. 
- 제적 위험자의 경우에는 `AttendanceSystem.calculateExpulsionCrews()` 메서드에서 수행하고 있습니다.

현재 로직이 "매개변수로 주어진 시간이, 기준으로 정해진 시간(분) + AttendanceStatus.boundary보다 늦었다면"를 만족하는 AttendanceStatus만 filter로 걸러낸 후 첫 번째 AttendanceStatus 객체를 반환하고 있습니다.
여기서 sorted를 넣은 이유는, 현재 enum의 순서가 출석, 지각, 결석 순이기 때문에 values()로 값을 가져온다면 가장 첫 번째가 출석이 오게됩니다. 
하지만, 매개변수로 들어온 시간이 만약 10시 31분이고 출석 시간은 10시라면 AttendanceStatus.ABSENCE를 반환해야 하지만 정렬하지 않는다면 AttendanceStatus.ATTENDANCE를 반환하게 됩니다.
- enum의 순서를 바꾸면 되지 않을까? 라고 생각했지만 enum의 순서때문에 로직이 잘못될 가능성을 배제하고 싶어서 이렇게 작성했습니다.

enum의 순서가 로직에 연관이 있는 것에 대해서 로키의 생각이 궁금합니다.🤔

Rok93 replied:
> enum의 순서를 바꾸면 되지 않을까? 라고 생각했지만 enum의 순서때문에 로직이 잘못될 가능성을 배제하고 싶어서 이렇게 작성했습니다. 

이 부분은 저도 부기의 생각과 동일하게 enum 순서에 의존적이면 안된다고 생각해요. 😃
enum 요소의 순서에 의존적인 코드는  짜는 사람은 인지하고있기 때문에 문제가 없을 수 있지만 다른 동료들이 해당 로직을 수정할 일이 있을 때는 이 부분을 인지하지 못한채 수정할 확률이 높다고 생각해요. 
이는 버그로 이어지겠죠. 😅

Rok93 replied:
Expulsion이 무슨 의미인가했더니.. 방출, 제명의 뜻이었군요...? 🙄 
`isExpulsionCrew` 는 내부적으로 `경고`, `면담`, `제적`인 경우를 모두 true로 반환하고있군요? 
네이밍이 실제 행위와 상이한 것 같아요. 이런 네이밍은 코드를 읽는데 혼동을 줄 수 있을 것 같아요.  🤔 

또한 여기서 sort는 아레와 같이 Comparator의 comparingInt 함수를 활용해봐도 좋을 것 같아요. 😃

```java
    .sorted(Comparator.comparingInt(AttendanceStatus::getBoundary).reversed())
```

Rok93 commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,110 @@
+
+    public boolean isAlreadyTodayAttendance(final String crewName) {
+        final Crew crew = this.findCrewByName(crewName);
+        return crew.existTodayAttendance(LocalDate.now().withYear(2024).withMonth(12));
```
existTodayAttendance 를 `isAlreadyAttendance` 로 네이밍해보면 어떨까요? 🤔

changuii replied:
네이밍의 일관성 때문일까요? 아니면 관용적으로 사용되는 용어가 isAlready일까요? 
로키의 생각이 궁금합니다! 🤔

Rok93 replied:
어떤 타입을 반환하는 냐에 따라 메서드 네이밍이 어느정도 관용적으로 사용하는 관습이 있는데요. 
[해당 글](https://tecoble.techcourse.co.kr/post/2020-04-26-Method-Naming/)을 참고해보시면 도움이되실 것 같습니다. 

물론 이런 네이밍적인 부분은 팀의 규칙이 더 중요하기 때문에 무조건 이렇게해야한다! 이런 것은 아닙니다.

changuii replied:
관용적인 부분에 대해서 아직 감이 부족한 것 같습니다. 좋은 글 감사합니다. :)

Rok93 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,123 @@
+        this.attendances = attendances;
+    }
+
+    public static Crew of(final String name, final LocalDate inputLocalDate) {
```
정적 팩토리 메서드의 입장에서 메서드 파라미터로 들어오는 값이 입력 값인지 알 필요가 있을까요? 
네이밍에서 input 이라는 워딩은 제외해도 될 것 같아요. 😃

changuii replied:
감사합니다.

Rok93 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,123 @@
+
+public class Crew implements Comparable<Crew> {
+    private final CrewName name;
+    private final List<Attendance> attendances;
```
AttenctSystem이 N개의 Crew를 상태로 가지고 각 Crew는 상태 값으로 attendances를 상태로 가지고 있군요. 🤔

AttenctSystem이 attendances를 가지고 각각의 attendance가 crew를 가지는 구조도 고민해볼 수 있을 것 같은데요. 

현재 주어진 요구사항을 다시한번 생각해보면서, 두 구조 중 어떤 방식이 더 좋은 선택인지 고민해보시면 좋을 것 같아요. 😃

changuii replied:
# 비교
### Crew -> Attendances 
1. 출석 확인 : crew의 이름으로 된, attendance들 중, 오늘 날짜의 attendance가 필요하거나 추가해야 합니다.
2. 출석 수정 : crew의 이름으로 된, attendance들 중, 주어진 날짜의 attendance를 가져와 새로운 시간으로 수정합니다.
3. 크루별 출석 기록 확인 : crew의 이름으로 된, attendances를 가져와 출력합니다.
4. 제적 위험자 확인 : 각 crew들이 제적 위험자인지 확인 후 가져와 출력합니다. 


### Attendance -> Crew
1. 출석 확인 : crew 이름이 일치하는 attendance들을 찾아 오늘 날짜의 attendance를 추가합니다.
2. 출석 수정 : crew 이름이 일치하는 attendance들을 찾고 주어진 날짜의 attendance를 찾아 수정합니다.
3. 크루별 출석 기록 확인 : attendance들 중, 주어진 crew이름을 가진 attendance를 반환합니다.
4. 제적 위험자 확인 : 각각의 crew에 대해 이름으로 attendance들의 그룹을 만들고 해당 크루가 제적 위험자인지 확인 후, 계산합니다.

제가 Crew -> Attendance로 만든 이유 (Crew의 수 N, Attendance의 수 M)
3. 크루별 출석 기록 확인에서, Crew -> Attendance의 경우에는 Crew만 찾으면 되지만 (N) Attendance -> Crew의 경우에는 모든 Attendance를 비교해야 합니다. (N * M)
4. 제적 위험자 확인의 경우에도 각각의 크루에 대해서 Attendance들이 분리되어 있기 때문에 N * M을 비교해서 찾아야 합니다.

로키의 의견이 AttendanceSystem이기 때문에 Crew가 아닌 Attendance가 밖에 존재하는 것이 이해하기 자연스럽기 때문인가요?
로키의 생각이 궁금합니다.

Rok93 replied:
저 또한 출석 미션은 익숙하지 않은 도메인이라 많은 가능성을 열어두고 코멘트를 드리고있는데요. (제가 제안드린 방식이 더 좋다는 의도로 말씀드린건 아니니 부기의 생각이 더 맞다고 생각하신다면 밀고나가셔도 전혀 문제 없습니다! 😃)

`AttendanceSystem이기 때문에 Crew가 아닌 Attendance가 밖에 존재하는 것이 이해하기 자연스럽기 때문인가요` << 말씀하신 이 내용도 약간 포함되는 내용이긴 합니다. 😃  AttendanceSystem이 Crew 보다는 Attendance들을 상태로 가지는 것이 조금 더 직관적이라고 생각했어요. 

각 요구사항을 수행하기위한 시간복잡도도 대략적으로 생각해보긴했었는데, 크루별로 출석일자가 그룹핑되어있는 `Crew -> Attendances`가 확실히 조금 더 유리해보이네요. 👍👍

changuii replied:
저도 로키가 말해주시고 나서 고민해보았는데 CrewSystem이 아닌 AttendanceSystem에 Crew가 있는 것이 어색하다고 느꼈어요
하지만, 효율적인 측면에서는 앞에 경우에서 처럼 Crew가 Attendance를 가지는게 더 자연스럽다고 느꼈어요

이 두 부분에 대해서 어떤 기준으로 나누어야 할까요? 
1. 비효율적이더라도 직관적인 코드를 작성해 가독성을 높이는 경우
2. 효율적이더라도 직관성이 부족한 코드를 작성해 가독성이 떨어지는 경우

로키의 생각과 이런 상황에서 어떤 것을 선택하시고 어떤 기준이 있으신지 궁금합니다. 🤔

Rok93 replied:
항상 고민되는 영역인데요. 
정해진 답은 없는 것 같아요. 

저는 개인적으로 서비스 혹은 해당 기능의 특징이나 추후 확장 가능성을 고려해서 정하는 편이에요. 
추후 크게 확장될 여지가 없다면 비효율적이라도 직관적인 면을 조금 더 볼 수도 있지만, 
확장성이 크다면 효율적인쪽에서 최대한 직관성을 챙겨보려고 노력하는 편입니다. << 이런 경우에 혼자서 좋은 답이 잘 안나오기 때문에 팀에 고민 포인트를 공유하고 설계 논의 등을 진행하게되는 것이죠. 😃

Rok93 commented on src/main/java/domain/CrewName.java:
```diff
@@ -0,0 +1,24 @@
+
+import java.util.Objects;
+
+public class CrewName implements Comparable<CrewName> {
```
원시값 포장 잘해주셨네요. 👍

CrewName은 [VO](https://tecoble.techcourse.co.kr/post/2020-06-11-value-object/)라고 볼 수 있지 않을까요?

changuii replied:
감사합니다. 처음보는 키워드라 학습해보겠습니다.!

Rok93 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public boolean equals(final Object o) {
```
[equals는 hashCode와 함께 재정의](https://tecoble.techcourse.co.kr/post/2020-07-29-equals-and-hashCode/)해주세요. 😉

Rok93 commented on src/main/java/domain/Attendance.java:
```diff
@@ -0,0 +1,73 @@
+        return isEmpty;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (o instanceof Attendance) {
+            final Attendance target = (Attendance) o;
+            return dateTime.toLocalDate().equals(target.dateTime.toLocalDate());
+        }
+
+        return false;
+    }
```
equals는 직접 정의해주신걸까요? 

IntelliJ는 자동으로 equals와 hashCode를 정의해주는데요. 
`⌘(command) + N` 단축키를 눌러서 쉽게 생성할 수 있어요. 😉

<img width="535" alt="image" src="https://github.com/user-attachments/assets/39ff1864-691c-428a-aa29-c79f6def6de9" />


```java
    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Attendance)) {
            return false;
        }
        Attendance that = (Attendance) o;
        return isEmpty == that.isEmpty && Objects.equals(dateTime, that.dateTime);
    }
```

changuii replied:
각 Attendance를 동등성 비교할때, 날짜마다 하나의 Attendance만 존재해야 했기 때문에, 커스터마이징해서 Date만 비교하도록 작성했는데, 확인해보니 intellij에서 생성해주는 기능에 필드 선택이 있었네요 감사합니다 :)

Rok93 replied:
저는 객체의 특정 상태값만으로 동등성을 비교하는 부분은 조금 직관성과 일관성이 떨어진다고 생각해요. 🤔 
이 부분을 일일이 코드를 모두 알고 있지 않으면 추후 문제가 생길 여지가 있다고 생각해요.

changuii replied:
확실히 직관성이 떨어진다고 생각합니다.
이렇게 작성했던 이유는 Date에 대한 Attendance가 하나만 존재해야 했기 때문이었습니다.
차라리 훨씬 직관적으로 Map을 활용하는 편이 더 좋을까요?

로키의 생각이 궁금합니다 ~!

Rok93 replied:
음... 부기의 현재 구현상에서는 어떻게 녹일지가 조금 고민되는 것 같아요.  🤔

현재 이 Attendence의 동등성 비교는 `Crew.addAttendance` 할 때만 사용되는걸까요? 
맞다면 동등성 로직은 제거하고   `isAlreadyAttendanced`와 같이 이미 출석을 했는지, 메서드를 하나 만들어서 해당 부분을 대체해보면 어떨까요? 🤔 

Map을 말씀하신 부분은 어디인지 바로 이해하기가 어려워서요. 😢
`AS IS`, `TO BE` 로직을 코멘트로 달아주셔도 좋을 것 같습니다.

changuii replied:
아하 제가 너무 갑자기 말해서 이해 못하셨을 것 같아요! 죄송해요

### AS IS
현재의 로직은 List<Attendance>로 관리하고 있기 때문에 사실상 Attendance는 리스트 내에서 중복될 가능성이 존재해요!
그래서 제가 해결한 방법은 equals()를 구현해서 
1. Attendance가 List에 존재하는지 여부를 검사할때,
2. Attendance를 추가할 때, 
3. Attendance를 수정할때(날짜에 해당하는 값 삭제 후 새로 등록) 

사용하고 있어요.

### TO BE
그래서 생각해본 방법은 List<Attendance> 대신 Map<LocalDate, Attendance>로 바꿔 사용하는 거에요 그럼 값이 중복될  일도 없고, Attendance의 equals()도 직관적으로 사용할 수 있다고 생각했어요

Rok93 replied:
> Attendance의 equals()도 직관적으로 사용할 수 있다고 생각했어요

Map의 key로 LocalDate를 관리하면 equals를 정의할 필요가 없다고 생각하는데, 제가 잘못이해한걸까요? 🤔

changuii replied:
맞아요! 
Map의 key를 LocalDate로 관리하면 equals()가 지금처럼 LocalDate()만 비교하는 대신, 정말 LocalTime과 isEmpty()를 비교해 '동등성 비교'로 직관적이라고 생각했어요

Rok93 reviewed (changes requested):
부기 안녕하세요.  출석 미션을 함께하게된 로키입니다. 🤗 
1단계 잘 구현해주셨어요. 👍👍 

몇몇 코멘트 남겨두었으니 확인해서 반영 부탁드릴게요. 🙏 

아무래도 1차 리뷰 마감기한에 맞추다보니 시간이 많이 부족하셨을거라 생각해요. 
이제 조금 더 여유가 생겼으니, 신경쓰지 못한 부분들은 다시한번 고민해보시면 좋을 것 같아요. 🙏

changuii reviewed (commented):
리뷰 자세하게 해주셔서 감사합니다.
추가로 미션 수행하면서 궁금했던점이 기억나 질문 드립니다.

> 제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다. 출석 상태가 같으면 닉네임으로 오름차순 정렬한다.

위의 요구사항을 지키기 위해 `AttendanceSystem.calculateExpulsionCrews()`의 stream에서 sorted()를 통해 정렬했습니다.
하지만, 정렬에 대한 로직을 정의해야했고, sorted()에 람다로 처음에 작성하려고 했으나 Crew 내부의 상태가 외부로 노출되어야만 했고 따라서 Crew가 Comparable을 구현하고 compareTo() 메서드를 Crew에 두었습니다. 
- 이 경우 앞에서의 문제였던 Crew의 상태가 외부로의 노출은 방지할 수 있었습니다.

하지만, 문득 생각해보니 새로운 정렬 요구사항이 생긴다면 새로운 기준으로 정렬을 해야하는데 이 경우에는 어떻게 해야할지 모르겠습니다.
로키의 생각이 궁금합니다!! 🤔

changuii committed:
refactor : 불필요한 줄바꿈을 제거하라

changuii committed:
fix(ExpulsionStatus): EXPULSION 오타를 수정하라

changuii committed:
fix(ExpulsionStatus): 필드의 private final을 추가하라

- 값이 변경되고, 외부에서 접근을 못하도록 변경하였다.

changuii reviewed (commented):
리뷰 감사합니다!!

changuii committed:
refactor : 관용적으로 사용하는 용어로 변경하라

- Operation -> UserCommand로 변경하였다.
- 유저가 잘못된 입력을 하면, 발생하는 예외에 메시지를 추가하였다.

changuii committed:
refactor(ExpulsionStatus): ExpulsionStatus의 stream 연산을 더 직관적으로 변경하라

- filter로직을 메서드로 분리하여 이해하기 쉽도록 하였다.
- 정렬 과정을 메서드로 분리하여 수행하도록 했다. (정렬한다는 의미 전달)

changuii committed:
refactor: 더 직관적인 메서드 명을 사용하라

changuii committed:
refactor: AttendanceSystem의 불필요한 코드를 지워라

Rok93 commented:
> (추가질문) 제적 위험자는 제적 대상자, 면담 대상자, 경고 대상자순으로 출력하며, 대상 항목별 정렬 순서는 지각을 결석으로 간주하여 내림차순한다. 출석 상태가 같으면 닉네임으로 오름차순 정렬한다.
위의 요구사항을 지키기 위해 AttendanceSystem.calculateExpulsionCrews()의 stream에서 sorted()를 통해 정렬했습니다.
하지만, 정렬에 대한 로직을 정의해야했고, sorted()에 람다로 처음에 작성하려고 했으나 Crew 내부의 상태가 외부로 노출되어야만 했고 따라서 Crew가 Comparable을 구현하고 compareTo() 메서드를 Crew에 두었습니다.
이 경우 앞에서의 문제였던 Crew의 상태가 외부로의 노출은 방지할 수 있었습니다.
하지만, 문득 생각해보니 새로운 정렬 요구사항이 생긴다면 새로운 기준으로 정렬을 해야하는데 이 경우에는 어떻게 해야할지 모르겠습니다.
로키의 생각이 궁금합니다!! 🤔

정렬의 기준에 대해서는 요구사항이 자주 바뀔 수 있다고 생각해요.
혹은 여러 정렬 기준이 생길 수도 있어요. (이름순, 생성 순, 출석일순 각각에 대한 오름차순, 내림차순) 
만약 여러 정렬 기준에 대해서 서버에서 제공을 해줘야하는 상황이된다면 Comparble 인터페이스로 구현하는 점은 조금 어려워질 것 같아요. 😅 
그렇기 때문에 저는 Comparable 인터페이스를 구현하는 방식보다는 정렬을 수행하는 위치에서 조건예따라 정렬을 변경하는 것이 좋다고 생각해요. 


>  새로운 정렬 요구사항이 생긴다면 새로운 기준으로 정렬을 해야하는데 이 경우에는 어떻게 해야할지 모르겠습니다.
로키의 생각이 궁금합니다!! 🤔

요약하자면 결국은 필요한 요구사항을 위한 정렬 기준을 `sorted()` 메서드의 파라미터로 전달해줌으로써 유동적으로 대응해줄 것 같아요. 🤔

changuii committed:
refactor : 날짜를 받아오는 로직을 외부로 분리하고, AttendanceSystem이 관리하도록 하라

changuii reviewed (commented):
감사합니다! 반영하겠습니다.

changuii commented:
> 요약하자면 결국은 필요한 요구사항을 위한 정렬 기준을 sorted() 메서드의 파라미터로 전달해줌으로써 유동적으로 대응해줄 것 같아요. 🤔

그렇다면 이 부분에 대해서도 view에게 값을 전달하기 위해 getter()를 사용하는 것처럼 어쩔 수 없이 사용해야 하는 부분일까요?

changuii committed:
refactor: enum 객체들을 constant 패키지로 이동시켜라

changuii committed:
refactor(AttendanceSystem): Crew를 외부로 노출시키지 말고 내부에서 연산하라

changuii committed:
refactor: Attendance를 찾지 않고, Update 로직시에 snap샷을 반환하라

changuii committed:
refactor(AttendanceController): AttendanceController의 중복 코드를 제거하라

changuii committed:
refactor: Controller에서 dto로 convert하는 로직을 분리하라

changuii committed:
refactor: 사용하지 않는 import를 제거하라

changuii committed:
refactor: 예외 메시지를 추가하라

changuii committed:
refactor: AttendanceStatus의 패키지를 constants로 옮겨라

changuii committed:
refactor(AttendanceStatus): 각 stream 처리의 의미를 나타내기 위해 메서드를 분리하라

changuii committed:
refactor(Attendance): 중복되는 생성자를 제거하라

changuii committed:
refactor(Crew): 깊은 복사 대신, 방어적 복사만 진행하라

changuii committed:
refactor(Crew): Comparable<Crew>를 구현하는 대신, compareByExpulsionCount를 만들어 외부에서 비교할때 사용하도록 하여라

- 외부에 값을 노출하지 않고 정렬 기준을 유연하게 가져갈 수 있도록 바꾸었다.

changuii committed:
refactor(AttendanceSystem): 부정적인 isNot을 사용하지 마라

- attendance가 가능한지 물어볼때 부정적으로 묻지 않고 isAttendanceDay로 물어보도록 변경하였다.

changuii committed:
refactor(AttendanceController): retryUntilCommandQuit의 while()문의 조건과 실행 로직을 분리하여 나타내라

changuii committed:
fix(AttendanceController): 출석할때, 이미 출석했다면 수정으로 안내하고 이름과 날짜는 입력받지 않도록 하라

changuii committed:
refactor(AttendanceController): 메서드 순서를 정리하라

changuii committed:
refactor(Crew): 정적 팩토리 메서드로 Attendance를 생성하라

changuii committed:
test(AttendanceTest): Attendance.of()에 대한 단위 테스트를 작성하라

changuii committed:
test(AttendanceTest): Attendance에 대한 단위 테스트를 작성하라

changuii committed:
refactor(Attendance): 사용하지 않는 복사 생성자를 제거하라

changuii committed:
test(CrewTest): 단위 테스트를 작성하라

changuii committed:
test(TodayDateTimeGenerator): 단위 테스트를 작성하라

changuii committed:
test(AnswerCommnad): 단위 테스트를 작성하라

changuii committed:
test(AttendanceStatusTest): 단위 테스트를 작성하라

Rok93 commented on src/main/java/domain/constants/AttendanceStatus.java:
```diff
@@ -0,0 +1,59 @@
+        return now.isAfter(todayBoundary.plusMinutes(attendanceStatus.matchTimeMinuteBoundary));
+    }
+
+    private static List<AttendanceStatus> sortedDescStatusForMatchingTimeMinuteBoundary() {
```
> 그렇다면 이 부분에 대해서도 view에게 값을 전달하기 위해 getter()를 사용하는 것처럼 어쩔 수 없이 사용해야 하는 부분일까요?

View에게 값을 전달하기위해서 getter 사용은 어쩔 수 없는 것이라 생각해요. 😃

changuii committed:
test(ExpulsionStatusTest): 단위 테스트를 작성하라

changuii committed:
test(UserCommandTest): 단위 테스트를 작성하라

Rok93 commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,145 @@
+    public static AttendanceSystem of(final List<String> data, final DateTimeGenerator dateTimeGenerator) {
+        final LocalDate today = dateTimeGenerator.generateDate();
+        final List<Crew> crews = data.stream()
+                .map(d -> d.split(DELIMITER)[0])
```
여기서 0, 1은 어떤 의미인지 한눈에 파악이 안되는데요. 
매직넘버를 상수로 처리해서 의도를 드러내줏면 좋을 것 같아요.

Rok93 commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,145 @@
+    private static final LocalDate CHRISTMAS_DAY = LocalDate.of(2024, 12, 25);
+    private static final String DELIMITER = ",";
+    private final List<Crew> crews;
+    private final DateTimeGenerator dateTimeGenerator;
```
아마도 [남겨주신 코멘트](https://github.com/woowacourse/java-attendance/pull/18#discussion_r1966411350)의 결과로 `DateTimeGenerator`를 주입받아서 날짜에 대한 처리를 처리하신 것으로 보여요. 

현재는 dateTimeGenerator 가 제공해주는 날짜에 너무 의존적이라고 생각해요. 🤔 

현재 출석도메인이 하루의 날짜에 대해서만 출석을 관리하고있는데, 추후 여러 날짜에 대해서 출석을 관리한다고 생각해봤을 땐 메서드로 날짜 정보도 입력을 받아주는게 좋다고 생각해요.

Rok93 commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,145 @@
+import java.util.Objects;
+
+public class AttendanceSystem {
+    private static final LocalDate CHRISTMAS_DAY = LocalDate.of(2024, 12, 25);
```
12월에는 다행히도 휴일이 하루 뿐이지만, 추후 12월 외에도 출석을 관리하게된다면, 이 휴일에 대한 상수들은 많이 늘어나게될 것 같아요. 

휴일 객체를 만들고 해당 객체에게 휴일 여부를 물어보면 어떨까요? 😃

Rok93 commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,145 @@
+
+    public boolean isAttendanceDay() {
+        final LocalDate today = dateTimeGenerator.generateDate();
+        return !(today.getDayOfWeek() == DayOfWeek.SUNDAY || today.getDayOfWeek() == DayOfWeek.SATURDAY
```
DayOfWeek에게 주말(= DayOfWeek.SUNDAY, DayOfWeek.SATURDAY)인지 여부를 메시지를 보내서 확인하면 참 좋겠지만, 
DayOfWeek는 저희가 만든게 아니기 때문에 새로운 책임을 부여하기에는 한계가있는데요. 🥲
이런 경우에는 별도의 Util 함수를 만들어볼 수도 있을 것 같아요. 

ex) **DateUtils.isWeekend(today)**

changuii committed:
test(AttendanceSystemTest): 단위 테스트를 작성하라

Rok93 commented on src/main/java/controller/AttendanceController.java:
```diff
@@ -0,0 +1,179 @@
+    private int inputDayOfMonthForUpdate(final String crewName, final AttendanceSystem attendanceSystem) {
+        outputView.printUpdateAttendanceDayOfMonth();
+        final int dayOfMonth = inputView.readDayOfMonth();
+        attendanceSystem.validateUpdateAttendanceDay(crewName, dayOfMonth);
```
여기서 처리를 하고있기보다는 실제로 업데이트하는 동작을 수행하기전에 체크하고 적절히 예외를 던져주는게 더 좋을 것 같아요.

changuii replied:
제가 이렇게 한 이유는 사용자가 수정을 위한 날짜를 입력받았을 경우, 수정할 수 없는 날짜라면 사용자에게 바로 피드백을 돌려줘 다시 날짜 입력을 받게하기 위함인데,
이 경우에는 로키가 알려준 VO로 만들어 해당 객체 내부에서 값을 검증하는게 좋을까요?

Rok93 replied:
네 저는 가능하면 객체의 생성에 대한 유효성 검증은 객체가 생성되는 시점에 검증하는 것이 좋다고생각해요. 😃

Rok93 commented on src/main/java/domain/AttendanceSystem.java:
```diff
@@ -0,0 +1,141 @@
+                .ifPresent(crew -> crew.updateAttendanceByDateTime(data[1]));
+    }
+
+    public Attendance attendance(final String name, final LocalTime time) {
```
AttendanceSystetm에대한 테스트가 없는 것 같은데, 테스트 코드 보강해주시면 좋을 것 같아요. 😃

그 외에도 누락된 테스트없는지 전체적으로 확인한번 부탁드릴게요. 🙏

Rok93 commented on src/main/java/domain/constants/ExpulsionStatus.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Comparator;
+import java.util.List;
+
+public enum ExpulsionStatus {
```
ExpulsionStatus, AttendanceStatus는 출석이라는 도메인에서 핵심적인 요소 중 하나라고 생각해요. 
그래서 저라면 domain 패키지에 위치시킬 것 같은데 부기는 constants 패키지에 위치시킨 이유가 궁금합니다.

changuii replied:
enum을 학습할때, enum의 등장 유래가 상수 클래스(클래스 내부에 상수로 관리)로 관리하던 상수들이 여러 문제가 존재해 등장했고, 때문에 저는 enum을 상수로 생각했습니다!

Rok93 replied:
enum class는 단순히 상수의 집합체로만 바라보기보다는,  비슷한 성질의 상수들을 모아서 만든 객체라고 생각해주시면 좋을 것 같아요. 😃

단순 상수였다면 enum 클래스는 별다른 역할이 없지 않았을 것 같은데, 부기가 구현해주신 enum class는 객체로써 자신의 책임을 다하고있는 것으로보이네요. 😉

Rok93 commented on src/main/java/domain/Crew.java:
```diff
@@ -0,0 +1,121 @@
+        return new Crew(name, attendances);
+    }
+
+    private static boolean canDateAttendance(final LocalDate today) {
+        if (today.getDayOfWeek() == DayOfWeek.SATURDAY || today.getDayOfWeek() == DayOfWeek.SUNDAY
+                || today.equals(CHRISTMAS_DATE)) {
+            return false;
+        }
+        return true;
+    }
```
아래와 같이 변경해볼 수 있을 것 같아요. 😃

```suggestion
    private static boolean canDateAttendance(final LocalDate today) {
        return !(today.getDayOfWeek() == DayOfWeek.SATURDAY || today.getDayOfWeek() == DayOfWeek.SUNDAY
                && today.equals(CHRISTMAS_DATE);
```

Rok93 replied:
메서드 파라미터로 들어오는 날짜가 꼭 오늘일 필요가 있나요?  
설령 오늘일지라도 Crew 객체의 입장에서 메서드 파라미터로 오늘 날짜가 들어오는지를 알 필요는 없을 것 같아요. 😃

Rok93 reviewed (changes requested):
부기 피드백 반영 잘해주셨어요.  👍👍 
테스트 코드 작성이 안된 부분들이 좀 보이는데, 이 부분 확인해서 테스트 추가 부탁드릴게요. 😉
추가적으로 몇몇 코멘트 남겨두었는데 확인 부탁드릴게요. 

다음 리뷰요청 때는 1단계는 이만 머지하도록할께요. 😃

changuii commented:
### 변경점
1. 받은 피드백에 대한 반영
   - Controller에서 AttendanceSystem을 멤버 변수가 아닌 지역 변수로 변경하였습니다. 
   - 굳이 멤버 변수로 둘 이유가 없음(에러 유발)
   - 미래에 확장될 가능성을 고려(데이터 정합성)

2. Operation을 Command로 변경 (관용적 표현)
   -  이해하기 쉽게 변경

3. 출석 수정에 대한 로직을 controller에서 분리하지 않고 AttendanceSystem이 담당
   -  모든 코드를 이상적인 객체지향으로 만들기는 현실적으로 힘듬 (어느정도의 타협이 필요)
   - 우리에게 매우매우 중요한 비즈니스 로직이 controller에서 수행됨 

4. AttendanceSystem 객체 생성을 Factory로 분리
   - 객체 생성 과정을 테스트 가능하도록 함
   - controller의 책임이 너무 많기 때문에 어느정도의 분리 필요

5. status들의 boundary를 좀 더 자세하게 작성
   - 코드는 요구사항을 잘 모르더라도 이해할 수 있는 수준으로 작성해야함

6. 정렬 로직을 이해하기 쉽게 변경
   - 앞에서도 말했듯이 이해하기 쉽게 해야하기 때문에 메서드를 통해서 정렬 로직을 분리하여 코드를 읽는 사람 입장에서 이해하기 쉽게 변경

7. existTodayAttendance를 isAlreadyAttendance 관용적인 표현으로 사람들이 더욱 쉽게 이해할 수 있도록 코드 작성

### 이외의 변경점

1. test 코드를 다시 작성
   -  변경된 부분이 많아 처음부터 새로 작성

2. 리팩터링 과정에서 생긴 불필요한 메서드, 매개변수 들을 제거

3. controller의 retryUntilCommandQuit의 로직을 수정
   -  while(x){ y } x 위치의 값은 비교하기 위한 값, y 위치에는 로직이 들어가도록 변경

4. isNot처럼 부정적인 용어 대신, 긍정적인 용어로 변경

5. Crew가 Attendances를 반환할때, 깊은 복사를 하고 있던 것을 방어적 복사로 변경
   - Attendance가 이미 불변객체이기 때문에 굳이 깊은 복사를 할 필요가 없다고 판단

6. Crew가 Comparable을 구현하는 것이 아닌, 메서드를 통해 비교 조건을 반환하도록 변경
   - 이를 통해 내부 상태를 외부로 노출하지 않고 특정 정렬 로직을 통해 외부에서 정렬할 수 있었다.

7. enum들을 domain.constants로 패키지 이동
   - 가독성 향상

8. controller에서 dto를 convert하는 로직을 Converter로 이동
   - controller의 책임이 너무 많다고 판단하여 이를 분리하였다.

9. 날짜를 가져오는 연산을 외부로 분리하였다.
   - 테스트의 용이성과 책임을 분리

changuii committed:
refactor(AttendanceSystem): TimeGenerator의 의존을 외부로 분리하라

- 외부에서 날짜를 받도록 변경하였다.

changuii committed:
refactor(AttendanceSystem): 휴일이 늘어날 것을 대비하여 휴일 enum을 만들어 분리하라

- 휴일을 물어보는 책임을 휴일 객체의 책임으로 분리하였다.
- 마찬가지로, 주말을 확인하는 책임도 DateUtils로 분리하였다. (자바에서 제공하는 클래스에 기능을 추가할 수 없었다.)

changuii committed:
test(AttendanceSystem): 테스트를 성공하도록 수정하라

- 변경되어 컴파일 오류나던 부분을 수정했다.

changuii committed:
refactor(AttendanceSystem): 매직 넘버를 상수로 처리하라

changuii committed:
test(Holiday): 단위 테스트를 작성하라

changuii committed:
test(AttendanceSystem): 실패 테스트를 작성하라

changuii committed:
refactor(Attendance): equals()를 정의할때, hashCode도 같이 정의하라

changuii reviewed (commented):
### 수정 사항
---
1. TimeGenerator를 Controller에서 관리하고 도메인의 메서드로 전달
AttendanceSystem의 TimeGenerator의 의존을 외부로 분리하라

2. 휴일이 늘어날 것을 대비해, 휴일객체에게 메시지를 던져 휴일 여부를 물어본다.
추가로 토요일과 일요일에 대해서도 같은 처리를 했다.

3. AttendanceSystem의 매직 넘버를 상수로 처리하도록 변경하였다.
4. 전체적인 테스트 코드 보강

Rok93 commented on src/main/java/domain/constants/Holiday.java:
```diff
@@ -0,0 +1,31 @@
+    Holiday(final LocalDate date) {
+        this.date = date;
+    }
+
```
불필요한 줄바꿈은 제거해주세요. 👀 

```suggestion
```

Rok93 commented on src/main/java/domain/constants/Holiday.java:
```diff
@@ -0,0 +1,31 @@
+                .isPresent();
+    }
+
+    private static boolean isEqualsMonthAndDayOfMonth(final LocalDate date, final LocalDate target) {
```
저도 잘 사용하지 않아서 이런게 있는지는 몰랐었는데, 역시 찾아보니 있네요. 
월,일 정보만 필요하다면 LocalDate 대신에 [MonthDay](https://lts0606.tistory.com/257)를 활용해보면 어떨까요?

Rok93 commented on src/test/java/domain/AttendanceSystemTest.java:
```diff
@@ -0,0 +1,211 @@
+                    .hasMessageContaining(ErrorMessage.CREW_NOT_FOUND.getMessage());
+        }
+
+        @DisplayName("수정할 수 없는 출석 일자라면, 예외가 발생한다.")
```
해당 테스트가 깨지는 것 같아요. 
확인 부탁드릴게요. 

<img width="808" alt="image" src="https://github.com/user-attachments/assets/987ea942-1a5a-4ff1-b925-14ac7d2d15bb" />

Rok93 reviewed (approved):
부기 피드백 반영 잘해주셨어요 👍💯
1단계는 이정도면 충분한 것 같아서 이만 머지하도록하겠습니다. 
추가로 깨지는 테스트 코드가 1개 있는데, 수정도 부탁드릴게요. 🙏
